### ISSUE TITLE
FP: unused_unit: closure lifetime binder

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#9883]
### Summary

.

### Lint Name

unused_unit

### Reproducer

I tried this code:

```rust
#![feature(closure_lifetime_binder)]

fn main() {
    let _ = for<'a> || -> () {
        let _: &'a bool = &true;
    };
}
```

I saw this happen:
` cargo clippy  --fix -- -Aclippy::all -Wclippy::unused_unit`
```
    Checking a v0.1.0 (/tmp/a)
warning: failed to automatically apply fixes suggested by rustc to crate `a`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error: implicit types in closure signatures are forbidden when `for<...>` is present
 --> src/main.rs:4:24
  |
4 |     let _ = for<'a> || {
  |             -------    ^
  |             |
  |             `for<...>` is here

error: aborting due to previous error

Original diagnostics will follow.

warning: unneeded unit return type
 --> src/main.rs:4:23
  |
4 |     let _ = for<'a> || -> () {
  |                       ^^^^^^ help: remove the `-> ()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unused_unit
  = note: requested on the command line with `-W clippy::unused-unit`

warning: `a` (bin "a") generated 1 warning (run `cargo fix --bin "a"` to apply 1 s
```

### Version

```text
rustc 1.67.0-nightly (b833ad56f 2022-11-18)
binary: rustc
commit-hash: b833ad56f46a0bbe0e8729512812a161e7dae28a
commit-date: 2022-11-18
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```

**Comment by y21** [comment]
There are no warnings in the reproducer anymore. #9849 probably fixed this?

**Comment by matthiaskrgr** [comment]
we should have a test than to make sure this doesn't regress at some point :)

**Comment by y21** [comment]
I think such a test was already added by the PR: https://github.com/rust-lang/rust-clippy/blob/0b63e95dce4e7f89936fa3547291bfe0b6ef5ffd/tests/ui/unused_unit.rs#L92-L96
or do you think your case is different that it also deserves its own test? It looks like the same issue to me

**Comment by y21** [comment]
This was fixed in #9849 , and the test that was added there seems basically equivalent to the repro here, so I think it can be safely closed
